README.md
tenvolt-chaser
==============

This project is a simple low cost programmable light chaser intended to be used 
as a component in a ten volt stage lighting system. It can control up to four
channels in a system by bumping the voltage up to ten volts in any pattern set
by the user. It has an onboard ten volt regulator and can be used as the sole 
controller in a system. 

Building and programming
------------------------

The code for the MSP430G2221 is written in C and intended to be compiled using
the msp430gcc compiler and tool set. The code uses the standard TI header files
so it may be compilable in Code Composer Studio, but this was not tested.

in addition, an rf2500 compatable MSP430 programming tool and mspdebug must be
used to program the chip. An MSP430G2 Launchpad was used in development.

To build, simply type: 
	
	make
	
To program, simply type

	make prog
	
Operation
---------

The chase pattern can be programmed using the bump switches on the board.
Press the program button to set the unit into program mode, then use the
bump switches to toggle the desired fixtures on or off. Pressing STARTSTOP
will save the pattern and advance to the next step. Repeat this process until
the desired pattern is programmed and then push the program button again.

The system can store up to 64 program steps and if this is reached, it will
save the pattern up to that many, then go back into execute mode.

In execute mode, the heartbeat led will pulse at the current tempo. This can
be modified by tapping the TAP button at the proper tempo. New settings take
effect immediately and can be changed on the fly.

In execute mode, the STARTSTOP button pauses and unpauses the chase routine,
flashing lights according to the program set in program mode.

Details
-------

The system requires at last 12vdc power to operate, however it is not 
recommended to provide much more than this, as heat may become an issue in 
the ten volt regulator as well as the 3.3 volt digital system voltage 
regulator.

There is a ten volt signal input and output on the board. Signals coming in
to the board from the input connector are ORed with the bump signals from the
microcontroller through a diode. There is about .2 to .7 volts loss depending
on the diode used which may affect the maximum light intensity avaliable to 
controllers before the chaser board, however this is most likely unnoticable.

There is an onboard ten volt regualtor which makes it possible to use this 
board as a standalone controller with full on or full off capability, 
however there is no provision for a blackout mode. 

The ten volt supply can be exported from the board should a user care to
implement other control methods such as varying control votlages with minimal
extra complexity.

Contact
-------

Contact Cirvinfox via github regarding any questions/issues/features. I would
love to hear from anyone wanting to make one of these themselves.


changelog
2014-12-5 CA Irvin
	* Added programmability
	* Added proper tap tempo function
	* Made it Actually Work (tm)

2014-11-28 Olivia Theze
	* Normalised code style and comments.
	* Moved function prototypes and defines to header instead
	* Added Vim modeline 
	* Fixed Changelog format
	* Fixed other style issues.

2012-10-26 CA Irvin
	* Removed sequence properties
	* Added serial routines



main.c
/* Corbin Irvin 2011
 
 Changelog:
 10/26/2012 - CA Irvin
 Removed sequence properties
 Added serial routines
 */
//#include <msp430g2221.h>  
#include <msp430.h>
#include <signal.h>
#include <inttypes.h>
#include "chaser.h"

/* global vars */
volatile uint32_t regTimer = 0;	// timer register, starts at 0 on startup, 500uS steps
volatile uint8_t	keyRegOld;
volatile uint8_t	keyReg;		// register that holds key state
volatile uint8_t	keyPulse;

volatile int		tapState =0;
volatile uint32_t delayTime = 500; // delay time in half milliseconds

void main(void)
{
	uint8_t keyreg = 0;
	int i =0;
	
	/* watchdog timer is pwm timebase */
	WDTCTL = WDTPW | WDT_MDLY_0_5;			// Set up watchdog to exp, 0.5ms
	IE1 |= WDTIE;							// enable wtd interrupt
  	
	/* configure the internal oscillator */
	BCSCTL1 = CALBC1_1MHZ_;	// set up for 1mhz stable operation without a crystal
	DCOCTL = CALDCO_1MHZ_;
  
	initIO();
	/* enable interrupts */
	//dint(); // opposite is 
	eint();
	
	/* switch to low power mode, act on interrupts */
	//LPM0;			
	
	while(1)
	{
		delayMs(500);
		_togglePort(2, LED_HEARTBEAT);
	
	}
}


/* 	
	this interrupt fires at 2KHz and does the pwm routine
*/
interrupt(WDT_VECTOR) WTD_interrupt(void)
{   
	uint32_t oldTime;
	// increment the time counter   	
	regTimer++;
	// update the keys
	keyRegOld = keyReg;
	keyReg = scanKeys();
	keyPulse = (keyReg ^ keyRegOld) & keyReg;

	
	// state machine for setting the LED tempo
	switch(tapState)
	{
		case 0:
		if(keyPulse & BIT5)
		{
			tapState = 1;
			oldTime = regTimer;
			_setPort(2, LED_HEARTBEAT);
		}
		break;
		
		case 1:
		if(keyPulse & BIT5)
		{
			tapState = 0;
			delayTime = regTimer - oldTime;
			_clearPort(2, LED_HEARTBEAT);
		}
		break;

		default:
		tapState = 0;
		break;
	}

	// Control the output channels, if a button is pressed, 
	// the channel must be bumped.
	if(keyReg & BIT3)
	{
		_setPort(1, EN_CH1);
	}
	else
	{
		_clearPort(1, EN_CH1);
	}
	if(keyReg & BIT0)
	{
		_setPort(1, EN_CH2);
	}
	else
	{
		_clearPort(1, EN_CH2);
	}
	if(keyReg & BIT4)
	{
		_setPort(1, EN_CH3);
	}
	else
	{
		_clearPort(1, EN_CH3);
	}
	if(keyReg & BIT1)
	{
		_setPort(1, EN_CH4);
	}
	else
	{
		_clearPort(1, EN_CH4);
	}


} 



/* initalizes the io ports */
void initIO(void)
{
	
	/* configure IO ports */
			
	P1SEL = 0x00;		// disable any special functions
	P2SEL = 0x00;
	
	P1DIR = 0x8F;		// set up the port directions
	P1REN = 0x70;		// enable internal resistors
	P1OUT = 0x70;		// set resistors to pullup
	
	P2DIR = BIT6 ;		// set up port directions
	P2REN = BIT7;		// enable internal resistors
	P2OUT = BIT7;		// set resistors to pullup
	
	return;
}

/*
	Scans the 6 input keys and updates a 8 bit value containing the button statuses
	BIT0	-	BTN2	-	CH2
	BIT1	-	BTN4	-	CH4
	BIT2	-	BTN6	-	STARTSTOP
	BIT3	-	BTN1	-	CH1
	BIT4	-	BTN3	-	CH3
	BIT5	-	BTN5	-	TAP
	BIT6	-	Reserved
	BIT7	-	Reserved
*/
uint8_t scanKeys(void)
{
	uint8_t keys=0;

	_clearPort(1, BTNC1);

	brief_pause(10000);

	keys = (P1IN & 0x70);

	keys = keys >> 3;

	_setPort(1, BTNC1);

	brief_pause(10000);

	keys |= (P1IN & 0x70);

	keys = keys >> 1;

	return  ((~keys)& 0x3f);
}

static void __inline__ brief_pause(register unsigned int n)
{
    __asm__ __volatile__ (
	"1: \n"
	" dec	%[n] \n"
	" jne	1b \n"
    : [n] "+r"(n));
}
	

void delayMs(uint32_t delay)
{
	uint32_t i ;
	i = regTimer + delay;
	while( regTimer < i ) nop();
	return;
}


chaser.c
/*
 * =====================================================================================
 *
 *       Filename:  chaser.c
 *
 *    Description: 
 *
 *        Changed: 2014-11-28
 *		
 *      Author(s):  Corbin Irvin (cirvinfox)
 *                  Olivia Theze (foxiepaws), fox@foxiepa.ws
 *
 * =====================================================================================
 */

/* 
 * (c) 2011 Corbin Irvin 
 *
 * Please see LICENSE for License details.
 *
 */

#include <msp430.h>
#include <signal.h>
#include <inttypes.h>
#include "chaser.h"

/* global vars */
volatile uint32_t   regTimer = 0;	 /* timer register, starts at 0 on startup, 500uS steps */
volatile uint8_t	keyRegOld;
volatile uint8_t	keyReg;          /* register that holds key state */
volatile uint8_t	keyPulse;

volatile int		tapState =0;
volatile int		heartbeatState;
volatile uint32_t   delayTime = 500; /* delay time in  milliseconds */
volatile uint32_t	heartbeatStartTime;
volatile uint32_t	oldTime;
volatile int		tapFlag = 0;

volatile uint8_t	programArray[MAXSTEPS] = { 0x01, 0x02, 0x04, 0x08};
volatile int		programStep = 0;
volatile int		programLength = 4;
volatile uint8_t	programState = 0;
volatile uint32_t	programCycleStartTime;

volatile int 		programmingState =0;

volatile int 		mode = 0; /* 0 is normal, 1 is programming */

void main(void)
{
	/* watchdog timer */
	WDTCTL = WDTPW | WDTHOLD;	/* WDT disabled */			
  	
	/* configure the internal oscillator */
	BCSCTL1 = RSEL3 + RSEL2 + RSEL1 + RSEL0;	/* set up for 15.25MHz internal oscillator */
	DCOCTL = DCO2 + DCO1 + DCO0;

	/* set up inputs and outputs */
	initIO();

	/* configure timerA0 for 1ms interrupt rate */
	TA0CTL = TASSEL_2 + MC_1 + TACLR ; /* SMCLK source, no division, up to TACCR0, interrupts on */
	TACCTL0 = CCIE;	/* enable the interrupt on compare */
	CCR0 = 15250 - 1;	/* count requried for a 1ms interrupt rate */


	/* enable interrupts */
	/* dint(); */ /* opposite is  */
	eint();
	
	/* switch to low power mode, act on interrupts */
	/* LPM0; */			
	
	while(1)
	{
		delayMs(500);
	
	}
}

/* 	
	this interrupt fires at 1KHz and does the pwm routine
*/
interrupt(TIMERA0_VECTOR) TimerA0_interrupt(void)
{   
	uint8_t temp;
	/* increment the time counter */   	
	regTimer++;
	/* update the keys */
	keyRegOld = keyReg;
	keyReg = scanKeys();
	keyPulse = (keyReg ^ keyRegOld) & keyReg;

	switch(mode)
	{
		case 0:
			if( (keyPulse & BIT6) && (tapState == 0) )
			{
				programStep=0;
				mode = 1; /* go to programming mode */
			}

			/* state machine for setting the LED tempo */
			switch(tapState)
			{
				case 0:
				 if(keyPulse & BIT5)
				 {
					 tapState = 1;
					 oldTime = regTimer;
				 }
				break;
				
				case 1:
				if(keyPulse & BIT5)
				{
					tapState = 0;
					delayTime = (regTimer - oldTime);
					heartbeatState = 0;
				}
				break;

				default:
				tapState = 0;
				break;
			}
	
			/* light sequence program state machine */
			switch ( programState ) 
			{
				case 0:	/* program not running */
					if(keyPulse & BIT2) /* if the startstop button is pressed, run prog */
					{
						programState = 1;
					}
					break;

				case 1:	/* program running, cycle start */
					if ( keyPulse & BIT2 ) /* turn off if we push the startstop button again */
					{
						programState = 0;
					}
					else
					{
						programCycleStartTime = regTimer;
						programState = 2;
					}
					break;

				case 2: /* program running, cycle step	 */
					if(keyPulse & BIT2) /* turn off if we push the startstop button again */
					{
						programState = 0;
					}
					else if(tapFlag)
					{
						/* time to go to the next step */
						/* write the program to the port */
						++programStep;
						if(programStep > programLength - 1)
						{
							programStep = 0;
						}
						programState = 1;
						tapFlag =0;
					}
					break;
					
				default:
					programState = 0;			
					break;
			}			
	

			/* Control the heartbeat */
			switch(heartbeatState)
			{
				case 0:
				_setPort(2, LED_HEARTBEAT);
				heartbeatState = 1;
				heartbeatStartTime = regTimer;
				tapFlag = 1;
				break;

				case 1:
				if(regTimer == heartbeatStartTime + 50)
				{
					_clearPort(2, LED_HEARTBEAT);
					heartbeatState = 2;
				}
				break;

				case 2:
				if(regTimer == heartbeatStartTime + delayTime)
				{
					heartbeatState = 0;
				}
				break;

				
				default:
					heartbeatState=0;

			}

			/* Control the output channels, if a button is pressed,
			 * the channel must be bumped. 
			*  |)}># */
			temp = (P1OUT & 0xF0);
			temp |= programArray[programStep];
			if(keyReg & BIT3)
			{
				temp |= EN_CH1;
			}
			if(keyReg & BIT0)
			{
				temp |= EN_CH2;
			}
			if(keyReg & BIT4)
			{
				temp|= EN_CH3;
			}
			if(keyReg & BIT1)
			{
				temp |= EN_CH4;
			}
			P1OUT = temp;
			
		break;

		case 1:
			/* Control the output channels, if a button is pressed,
			 * the channel must be bumped. 
			 */
			temp = (P1OUT );
			if(keyPulse & BIT3)
			{
				temp ^= EN_CH1;
			}
			if(keyPulse & BIT0)
			{
				temp ^= EN_CH2;
			}
			if(keyPulse & BIT4)
			{
				temp ^= EN_CH3;
			}
			if(keyPulse & BIT1)
			{
				temp ^= EN_CH4;
			}
			P1OUT = temp;
			
			if(keyPulse & BIT2) /* log the state of the lamps when STARTSTOP pressed */
			{
				programArray[programStep] = temp & 0x0F;
				++programStep;
				P1OUT &= 0xF0;
			}
			if((programStep > MAXSTEPS - 1) || (keyPulse & BIT6)) 
			{
				/* program mode exit, save the pattern */
				programLength = programStep;
				programStep = 0;
				mode = 0;
				programState = 0;
				heartbeatState = 0;
				tapState   = 0;
			}
		break;

		default:
			mode = 0;


	}

} 



/* initalizes the io ports */
void initIO(void)
{
	
	/* configure IO ports */
			
	P1SEL = 0x00;		/* disable any special functions */
	P2SEL = 0x00;
	
	P1DIR = 0x8F;		/* set up the port directions */
	P1REN = 0x70;		/* enable internal resistors */
	P1OUT = 0x70;		/* set resistors to pullup */
	
	P2DIR = BIT6 ;		/* set up port directions */
	P2REN = BIT7;		/* enable internal resistors */
	P2OUT = BIT7;		/* set resistors to pullup */
	
	return;
}

/*
 *  Scans the 6 input keys and updates a 8 bit value containing the button statuses
 *  BIT0	-	BTN2	-	CH2
 *  BIT1	-	BTN4	-	CH4
 *  BIT2	-	BTN6	-	STARTSTOP
 *  BIT3	-	BTN1	-	CH1
 *  BIT4	-	BTN3	-	CH3
 *  BIT5	-	BTN5	-	TAP
 *  BIT6	-	Reserved
 *  BIT7	-	Reserved
*/
uint8_t scanKeys(void)
{
	uint8_t keys=0;
	_clearPort(1, BTNC1);
	brief_pause(50);
	keys = (P1IN & 0x70);
	keys = keys >> 3;
	_setPort(1, BTNC1);
	brief_pause(50);
	keys |= (P1IN & 0x70);
	keys = keys >> 1;
	keys =  ((~keys)& 0x3f);
	if((P2IN & BIT7) == 0) 
	{
		keys |= BIT6;
	}
	return keys;
}

static void __inline__ brief_pause(register unsigned int n)
{
	 __asm__ __volatile__ (
	 "1: \n"
	 " dec	%[n] \n"
	 " jne	1b \n"
	 : [n] "+r"(n));
}
	

void delayMs(uint32_t delay)
{
	uint32_t i ;
	i = regTimer + delay;
	while( regTimer < i ) nop();
	return;
}

/* vim: set noet :*/


chaser.h
/*
 * =====================================================================================
 *
 *       Filename:  chaser.h
 *
 *    Description:  header file for chaser.c
 *
 *        Changed: 2014-11-28
 *
 *         Author:  Corbin Irvin (cirvinfox)
 *                  Olivia Theze (foxiepaws), fox@foxiepa.ws
 *
 * =====================================================================================
 */

/* Function Prototypes */
uint8_t scanKeys(void);
void initIO(void);
static void __inline__ brief_pause(register unsigned int n);
void delayMs(uint32_t);

/* Definitions */
#define MODESW_MASK 0xC0
/* MAXSTEPS defines the maximum pattern steps a user program can have */
#define MAXSTEPS 64

/* port 2 definitions */
#define LED_HEARTBEAT BIT6
#define BTN_PROGRA	BIT8
/* port 1 definitions */
#define EN_CH1	BIT0
#define EN_CH2	BIT1
#define EN_CH3	BIT2
#define EN_CH4	BIT3
#define BTNR1	BIT4
#define BTNR2	BIT5
#define BTNR3	BIT6
#define BTNC1	0x80;
/* const int BTNC1 = BIT7; */

/* macro to help us toggle bits on ports */
#define _togglePort(port,bit) P##port##OUT^= bit
#define _setPort(port,bit) P##port##OUT |= bit
#define _clearPort(port,bit) P##port##OUT &= ~bit

/* general bit manipulations */
#define _toggleBit(reg,bit) reg ^= bit  
#define _setBit(reg,bit) reg |= bit
#define _clearBit(reg,bit) reg &= ~##bit

